### PR TITLE
Extdev fpgas

### PR DESCRIPTION
**Pull request by rowleya**
Addresses changes in SpiNNakerManchester/PACMAN#463 including:
- Change how fixed keys are specified
- Add support for 2D (and beyond) vertices
- Change how command senders are added

### CODE REVIEW

File: spinn_front_end_common/utility_models/reverse_ip_tag_multi_cast_source.py [hunk 1]
```diff
@@ -226,5 +209,34 @@ def create_machine_vertex(
             assert (sdram == machine_vertex.sdram_required)
         return machine_vertex
 
+    def __filtered_send_buffer_times(self, vertex_slice):
+        ids = vertex_slice.get_raster_ids(self.atoms_shape)
+        send_buffer_times = self._send_buffer_times
+        n_buffer_times = 0
+        if send_buffer_times is not None:
+            # If there is at least one array element, and that element is
+            # itself and array
```
Comment (andrewgait) [inline spinn_front_end_common/utility_models/reverse_ip_tag_multi_cast_source.py:218]: "an array"

File: spinn_front_end_common/interface/interface_functions/command_sender_adder.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
+# Copyright (c) 2017-2019 The University of Manchester
```
Comment (Christian-B) [inline spinn_front_end_common/interface/interface_functions/command_sender_adder.py:1]: 2022

File: spinn_front_end_common/interface/interface_functions/command_sender_adder.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    def add_command_senders(self):
+        progress = ProgressBar(FecDataView.get_n_vertices(), "Adding commands")
+        for vertex in progress.over(FecDataView.iterate_vertices()):
+            if isinstance(vertex, AbstractSendMeMulticastCommandsVertex):
+                machine = FecDataView.get_machine()
+                link_data = None
+
+                # See if we need a specific placement for a device
+                if isinstance(vertex, ApplicationFPGAVertex):
```
Comment (Christian-B) [inline spinn_front_end_common/interface/interface_functions/command_sender_adder.py:55]: Consider moving class specific code to the classes

File: spinn_front_end_common/interface/interface_functions/command_sender_adder.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        if command_sender is None:
+            command_sender = self.__new_command_sender()
+            self.__command_sender_for_chip[(x, y)] = command_sender
+            cores = self.__cores(x, y)
+            p = cores[self.__system_placements.n_placements_on_chip(x, y)]
+            self.__system_placements.add_placement(
+                Placement(command_sender.machine_vertex, x, y, p))
+        return command_sender
+
+    def __new_command_sender(self):
```
Comment (Christian-B) [inline spinn_front_end_common/interface/interface_functions/command_sender_adder.py:108]: I would do this inline so you can add the xy to the label where applicable

File: spinn_front_end_common/utilities/database/database_writer.py [hunk 5]
```diff
@@ -140,6 +141,9 @@ def add_application_vertices(self):
         with self.transaction() as cur:
             # add vertices
             for vertex in FecDataView.iterate_vertices():
+                max_atoms = vertex.get_max_atoms_per_core()
+                if not isinstance(max_atoms, int):
```
Comment (Christian-B) [inline spinn_front_end_common/utilities/database/database_writer.py:145]: if type conversion is needed I would put it in the vertex so it works everywhere.

File: spinn_front_end_common/utilities/report_functions/reports.py [hunk 6]
```diff
@@ -267,9 +267,7 @@ def _write_one_vertex_partition(f, vertex):
     machine_vertices = sorted(machine_vertices,
                               key=lambda x: x.vertex_slice.lo_atom)
     for sv in machine_vertices:
-        f.write("  Slice {}:{} ({} atoms) \n".format(
-            sv.vertex_slice.lo_atom, sv.vertex_slice.hi_atom,
-            sv.vertex_slice.n_atoms))
+        f.write("  Slice {}\n".format(sv.vertex_slice))
```
Comment (Christian-B) [inline spinn_front_end_common/utilities/report_functions/reports.py:270]: f strings

**Review by Christian-B — APPROVED** [review]
a few comments none which can not wait for version2

File: spinn_front_end_common/interface/interface_functions/command_sender_adder.py (hunk 2 above)
Comment (rowleya) [reply spinn_front_end_common/interface/interface_functions/command_sender_adder.py]: Fixed

File: spinn_front_end_common/interface/interface_functions/command_sender_adder.py (hunk 3 above)
Comment (rowleya) [reply spinn_front_end_common/interface/interface_functions/command_sender_adder.py]: This has been moved now

File: spinn_front_end_common/interface/interface_functions/command_sender_adder.py (hunk 4 above)
Comment (rowleya) [reply spinn_front_end_common/interface/interface_functions/command_sender_adder.py]: I added the label as an argument

File: spinn_front_end_common/utilities/database/database_writer.py (hunk 5 above)
Comment (rowleya) [reply spinn_front_end_common/utilities/database/database_writer.py]: These lines have been deleted, but the functionality of the maximum atoms per core has been refactored to account for this anyway

File: spinn_front_end_common/utilities/report_functions/reports.py (hunk 6 above)
Comment (rowleya) [reply spinn_front_end_common/utilities/report_functions/reports.py]: Fixed

**Pull request merged**